### ISSUE TITLE
/7-tips-to-handle-undefined-in-javascript/

### ISSUE THREAD

**Issue by panzerdp** [panzerdp/dmitripavlutin.com-comments#29]
Written on 02/07/2017 11:30:19 

URL: https://dmitripavlutin.com/7-tips-to-handle-undefined-in-javascript/

**Comment by panzerdp** [comment]
> Comment written by **Kutsan Kaplan** on **04/17/2017 10:05:06**

<p>About semicolon; I think as long as you didn't start a line with <code>(</code>, <code>[</code>, or <code>`</code>, it's okay.</p><p>By the way, that was great read. I learned a lot of from you. Thank you.</p>

**Comment by panzerdp** [comment]
> Comment written by **Dmitri Pavlutin** on **04/17/2017 10:57:46**

<p>Thanks @Kutsan:disqus! I'm glad you learned a lot from the article.</p>

**Comment by panzerdp** [comment]
> Comment written by **Eugene Karataev** on **04/18/2017 16:15:16**

<p>"Call Object.assign({ }, unsafeOptions, defaults) to build a new object options."<br>Looks like a typo - 2nd and 3rd arguments should be swapped.</p>

**Comment by panzerdp** [comment]
> Comment written by **Dmitri Pavlutin** on **04/18/2017 16:55:19**

<p>Nice catch @eugenekarataev:disqus. Fixed.</p>

**Comment by panzerdp** [comment]
> Comment written by **Dmitry Manannikov** on **04/18/2017 17:48:07**

<p>FYI <br>Object.assign works differently in Safari when user with DOMStringMap. Like Object.assign({}, element.dataset)</p>

**Comment by panzerdp** [comment]
> Comment written by **PhistucK** on **04/22/2017 10:12:39**

<p>Might have been fixed in the latest Safari release (10.1? iOS 10.3? I forget).<br><a href="https://bugs.webkit.org/show_bug.cgi?id=164114" rel="nofollow noopener" title="https://bugs.webkit.org/show_bug.cgi?id=164114">https://bugs.webkit.org/sho...</a></p>

**Comment by panzerdp** [comment]
> Comment written by **nnnnnn321** on **05/01/2017 02:54:36**

<p>I found the <code>append(array, config)</code> function a little confusing at first glance, because initially I assumed the <code>config</code> argument held configuration information, but it really doesn't.</p><p>Your comments about semicolon usage in the context of the <code>return</code> statement don't really make sense, because your suggested fix works with or without a semicolon. Yes, ASI is what causes the problem, but no, always using semicolons doesn't prevent it, you still need to know to put something on the same line as the <code>return</code> keyword.</p>

**Comment by panzerdp** [comment]
> Comment written by **Lorraine Lee** on **05/02/2017 15:34:49**

<p>"By itself accessing a non-existing property does not throw an error. The real problem appears when trying to get data from a non-existing property value. This is the most common undefined related trap, reflected in the well known error message TypeError: Cannot read property &lt;prop&gt; of undefined."</p><p>Oh, wow, I get it now (I think).  Thank you eversomuch!  Basically (if I grok this right), undefined works like property insurance.  Without property insurance, you're one catastrophic event away from financial ruin.  With property insurance you're two catastrophic events away from financial ruin.</p>

**Comment by panzerdp** [comment]
> Comment written by **Dmitri Pavlutin** on **05/02/2017 18:16:39**

<p>Hey @Tharana:disqus,<br>That's interesting. I agree, undefined in a way is an insurance.</p>

**Comment by panzerdp** [comment]
> Comment written by **stephband** on **05/04/2017 17:51:43**

<p>Great article!</p><p>My only problem with using the 'in' operator for detecting the existence of a property, is that if you have set a property to undefined, the 'in' operator will return true...</p><p>    const a = { x: undefined };<br>    'x' in a;     // true</p><p>    const b = {};<br>    'x' in b      // false</p><p>Now arguably you should use the 'delete' operator if you are deleting properties from an object (although it is slower than setting to undefined in V8), but even so, it's very easy to assign undefined to an object property. So I far prefer the explicit equality check for property detection...</p><p>    const a = { x: undefined };<br>    a.x !== undefined;    // false</p>

**Comment by panzerdp** [comment]
> Comment written by **Dmitri Pavlutin** on **05/05/2017 05:36:02**

<p>Thanks @stephband:disqus!<br>I think assigning `undefined` to a property doesn't make much sense. When a property is not necessary, just don't create it.<br>Otherwise thanks for describing the edge situation about `in` operator.</p>

**Comment by panzerdp** [comment]
> Comment written by **Ricardo Sánchez** on **05/11/2017 07:08:57**

<p>My favorite way to deal with javascript undefined/null is to use purescript.<br>Maybe your tip number 8 should be the usage of languages that compile to js.</p>

**Comment by panzerdp** [comment]
> Comment written by **Dmitri Pavlutin** on **05/17/2017 16:07:21**

<p>If you follow the 7 recommendations listed, you won't have any undefined related issues.</p>

**Comment by panzerdp** [comment]
> Comment written by **Ricardo Sánchez** on **05/18/2017 06:50:08**

<p>3 4 5 6 are ways to hide the truth, it is just silently error that will become hard to debug.<br>Consider a super simple function called add(Num, Num), and for some reason the programmer (maybe from user data, or from json) has an object const o = {x:1}, and by definition he has to call the function with add(o.x, o.y), what is the correct default value? and why? 0? 1? why not -3? None, it is data corruption and it should be notified.<br>The programmer can assume (and it should be correct but it is not) that his function only expects Numbers.<br>It is better to have an fixable error than a working app with an unexpected behaviour.</p>

**Comment by panzerdp** [comment]
> Comment written by **hgoebl** on **05/19/2017 07:28:41**

<p>In my experience handling `undefined` and `null` as almost the same and comparing like `a == null` or `a != null` in contrast to `===` and `!==` is an easy way to go. Short code, less errors, just one more rule for the linter to accept `!=` and `==` in this special case when comparing null.<br>Who really cares whether something is null oder undefined? It's practically the same. The only difference is to detect non-provided arguments. But for this it's best to let Babel or TypeScript generate the detection code and use defaults and composition.</p>

**Comment by panzerdp** [comment]
> Comment written by **Dmitri Pavlutin** on **05/26/2017 10:39:49**

<p>Hi @hgoebl:disqus,<br>The example you describe a == null is probably a single recommended case of == operator usage.</p><p>"Who really cares whether something is null or undefined? It's practically the same." Generally yes. But practically the mechanism of default parameters reacts only to `undefined` values, which makes the distinction between undefined and null important.</p><p>Another moment is to understand the meaning of the two values. undefined stands for "uninitialized" state, while null stands for "missing object". Applying this distinction consistently makes the code more predictable at edge scenarios. I never allow undefined to spread within application data flow, but allow null to exist at one level of stack call to identify a missing object.</p>

**Comment by panzerdp** [comment]
> Comment written by **hgoebl** on **05/26/2017 10:43:01**

<p>I'm with you. Thanks for your post, BTW :-)</p>

**Comment by panzerdp** [comment]
> Comment written by **Dmitri Pavlutin** on **05/26/2017 10:51:45**

<p>Thanks! I'm glad you like it.</p>

**Comment by panzerdp** [comment]
> Comment written by **Buckwheat** on **05/28/2017 08:59:44**

<p>There's one strange way to handle undefined in functions, that I meted sometimes.<br>`function foo(a, b, undefined) {<br>if (a !== undefined  &amp;&amp; b !== undefined)<br>//..<br>}`</p>

**Comment by panzerdp** [comment]
> Comment written by **Dmitri Pavlutin** on **05/28/2017 16:41:55**

<p>Haha, that's a good one! Probably the weirdest JavaScript code I've seen. :)</p>

**Comment by panzerdp** [comment]
> Comment written by **Gorjan Mishevski** on **02/02/2018 13:24:48**

<p>Very nice article! Thanks :)</p>

**Comment by panzerdp** [comment]
> Comment written by **Dmitri Pavlutin** on **02/02/2018 17:48:17**

<p>You're welcome!</p>

**Comment by panzerdp** [comment]
> Comment written by **Yang NC** on **02/05/2018 12:18:23**

<p>Really nice article.<br>do you think  the default parameter and initial value for the variables should have a really business meaning?</p>

**Comment by panzerdp** [comment]
> Comment written by **Renzo** on **05/09/2018 02:43:57**

<p>Very nice article! Was wondering if const variable = object.prop || 'default value' is also a good idea.</p>

**Comment by panzerdp** [comment]
> Comment written by **Dmitri Pavlutin** on **05/10/2018 04:13:34**

<p>Thanks @disqus_ks9MgkrBHc:disqus!<br>const variable = object.prop || 'default value' doesn't work well when object.prop has a payload falsy value like false, 0, '' (empty string).</p>

**Comment by panzerdp** [comment]
> Comment written by **Josh Schneider** on **06/14/2018 19:03:32**

<p>This is an excellent post, thanks! Here's a probably dumb question, but I'm always finding myself unsure about the most fault-proof approach to testing for defined/undefined in `if` conditions (in blocks, ternaries, or similar).</p><p> Let's say I know that `obj.prop` is not possibly a falsy value. Is it fine in all semi-modern browsers (and in Node) to just do `if (obj.prop) {}` or `obj.prop ? ...` or `obj.prop &amp;&amp;`? Or should I be considering `if ('prop' in obj) {}` or even `if (typeof obj.prop !== 'undefined'){}`?</p><p>That last way used to be my goto method, since I think it was always safe, but for a long time now I've moved to terser approaches like `if (obj.prop) {}` ... again, assuming I'm sure that `obj.prop` won't be a falsy value, only defined or undefined. Also, I'm not bothering to coerce anything with `!!`, which is something I also wonder about (and don't particularly like doing).</p><p>EDIT: I just realized this is very close to @Renzo's question, but with more detail. Maybe I should just assume that knowing that `obj.prop` will never be falsy is the key, but any other thoughts are welcome.</p><p>Thanks</p>

**Comment by panzerdp** [comment]
> Comment written by **Dmitri Pavlutin** on **06/17/2018 15:43:19**

<p>I find `if ('prop' in obj) {}` syntax the most expressive and efficient to verify the property existence.</p>

**Comment by panzerdp** [comment]
> Comment written by **Josh Schneider** on **06/18/2018 13:56:30**

<p>Thanks. And assuming you know that `obj` will be either an object or undefined, would your condition be `if (obj &amp;&amp; 'prop' in obj) {}`?</p>

**Comment by panzerdp** [comment]
> Comment written by **Dmitri Pavlutin** on **06/20/2018 11:10:48**

<p>Yes, such verification could work.<br>But generally I don't recommend working with variables that might have `undefined` value.</p>

**Comment by panzerdp** [comment]
> Comment written by **Josh Schneider** on **06/20/2018 23:28:11**

<p>I tend to get overzealous with vanilla JS, in order to avoid undefined errors. I guess I was also just revisiting the question of needing to coerce anything in conditional tests. In this case, it's true, I should likely just know that 'obj' will exist. Thanks for the responses.</p>

**Comment by panzerdp** [comment]
> Comment written by **threecee** on **07/17/2018 14:28:52**

<p>Just found this very helpful post. Thanks!</p><p>I'm a beginner/intermediate in Javascript/node. Wish I'd found this earlier. It would have saved me countless hours of debugging!</p>

**Comment by panzerdp** [comment]
> Comment written by **Ron Royston** on **10/22/2018 16:18:26**

<p>Great article Dmitri! Thank you!</p>

**Comment by panzerdp** [comment]
> Comment written by **Balaji hope tutors** on **11/14/2018 12:28:56**

<p>Hope tutors is the leading <a href="https://www.hopetutors.com/course/java-training-in-chennai/" rel="nofollow noopener" title="https://www.hopetutors.com/course/java-training-in-chennai/">Java Online Course</a> provider in Chennai with 100% Placement Assurance.</p>

**Comment by panzerdp** [comment]
> Comment written by **Dmitri Pavlutin** on **07/05/2019 16:55:12**

<p>You're welcome!</p>

**Comment by panzerdp** [comment]
> Comment written by **Dmitri Pavlutin** on **07/05/2019 16:55:51**

<p>I'm glad my articles help advance your JavaScript knowledge. Subscribe to my newsletter to get new ones.</p>

**Comment by panzerdp** [comment]
> Comment written by **Dmitri Pavlutin** on **07/05/2019 16:57:17**

<p>I find default parameters useful when most of the times the parameter takes a specific, default value, and rarely other values.</p>

**Comment by panzerdp** [comment]
> Comment written by **yuk10z** on **05/05/2020 01:16:37**

<p><a href="https://designe.com.br" rel="nofollow noopener" title="https://designe.com.br">Designe</a></p>

**Comment by panzerdp** [comment]
> Comment written by **luisCarlos** on **10/12/2020 14:21:49**

<p>Hey Dimitri, i'm glad to have found your article and even though is over 2 years old, it did help me resolve and issued with nested arrays inside and object read from mongoDB. It kept giving me the undefined error on the embedded array but I could access all fo the non-array props.</p><p>it drove me nuts for several weeks until I found your article. when I tried the destructuring method with a default value assignment, then I found out why it was failing, the array was returning an undefined value on the first go around but not afterwards in React and once I put the default value I was able to use the map instruction to get at the array values....</p><p>PRETTY NEAT, Thank you :) <a href="https://media2.giphy.com/media/14sy6VGAd4BdKM/giphy.gif" rel="nofollow noopener" title="https://media2.giphy.com/media/14sy6VGAd4BdKM/giphy.gif">https://media2.giphy.com/me...</a></p>

**Comment by panzerdp** [comment]
> Comment written by **Dmitri Pavlutin** on **10/13/2020 07:51:43**

<p>That's great! I'm glad that my post has helped you solve the problem.</p>

**Comment by phutngo** [comment]
Wow this is the highest quality explanation on this topic! Thank you!

**Comment by panzerdp** [comment]
> Wow this is the highest quality explanation on this topic! Thank you!

You're welcome @phutngo!

**Comment by gutomezencio** [comment]
I never saw an good explanation regarding this topic like this one! Thank you so much for that!

**Comment by romina-trazzi** [comment]
As a newbie, I really love this article.

**Comment by panzerdp** [comment]
@romina-trazzi Glad you like it! 😎

**Comment by mdasifk302** [comment]
https://allbin.site/

**Comment by gunslingor** [comment]
A++ Article

One subject missed: The order of precedence of where each item should be fixed. Null/Undefined safety is actually a full stack issue, higher up the stack you go less you have to do and the lower the you go through they layers addressing the issue the more you have to do (e.g. "user.avatar || default" avatar could be in 100 places on the lowest layer of the view, 1 place on the highest layer of the ajax response layer, or 0 places entirely if the DB interface guarantees avatar is required, defaulted to something).  

Of course, Typescript is better... it basically just guides these decisions where the freedom of JS gives programmers too many options to work with... architecture/engineering, all about shrinking infinities.

**Comment by punund** [comment]
Tip 7 is plain wrong and misleading. Nowhere one MUST use a semicolon, except for the for(;;) body. People confuse semicolon insertion with "restricted production" rules.

This would work perfectly fine in any context:

"let a = 42
'foo'
return a"

unless you begin next line with something that can be a continuation, such as

"a = b
[c] = 42", but that's another story and has nothing to do with undefined.

One may get a wrong impression of correctness if you write

"return
   returnValue;"

because you used semicolon. However, this still returns undefined and has nothing to do with semicolons. In your example,

"return [ 
    2, 3, 5, 7, 11, 13, 17 
  ]"

works just fine with no semicolon.

**Comment by panzerdp** [comment]
> Tip 7 is plain wrong and misleading. Nowhere one MUST use a semicolon, except for the for(;;) body.

@punund My tip is correct and to the point. You just didn't get the idea: don't trust the automatic semicolon insertion until you understand how it works. If you insert a newline inside a restricted production (`return \n 42`), then the automatic semicolon insertion would create 2 statements (`return; 42;`) that behave differently than the original intent of returning 42.

By JS spec certain statements must be terminated with semicolons (ASI just inserts them for you when you don't use semicolons). This is what is [written](https://262.ecma-international.org/6.0/index.html#sec-automatic-semicolon-insertion) in specification.

> Certain ECMAScript statements (empty statement, let, const, import, and export declarations, variable statement, expression statement, debugger statement, continue statement, break statement, return statement, and throw statement) **must be terminated with semicolons**. Such semicolons may always appear explicitly in the source text. 

I also mention that you may omit them:

> What happens when you don't want to indicate these semicolons? In such a situation ECMAScript provides an [Automatic Semicolon Insertion](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-automatic-semicolon-insertion) (ASI) mechanism, which inserts for you the missing semicolons.

> Helped by ASI, you can remove the semicolons from the previous example:

But in the end ASI is going to insert the semicolons for you, but still they MUST be as per spec.

**Comment by punund** [comment]
You are correct about quoting the ECMA specs, which mention semicolons, automatically inserted or otherwise.

However, when you write
>If you use one of the above statements, be sure to indicate a semicolon at the end:

it directly contradics the specs, which uses the word "indicate" to denote a manully inserted explicit semicolon symbol in your code, while the spec you quote say:

>  _MAY_ always appear explicitly in the source text.

which impiles that they as well may not appear there.  I believe that furthering FUD about semicolons should not belong in this otherwise helpful article, as there exist many people who adhere to semicolonless style in JavaScript in their own right.

You may want to write another article about semicolons, since this topic, as you rightly mention, is rarely well understood.

**Comment by panzerdp** [comment]
@punund 
No, there is no contradiction. The issue is that you don't like my choice of words to explain what the spec says.

Corresponding to this part:

> MAY always appear explicitly in the source text.

I make the following explanation further in the article, which, for some reason, you don't mention at all in your comment:

> What happens when you don't want to indicate these semicolons? In such a situation ECMAScript provides an [Automatic Semicolon Insertion](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-automatic-semicolon-insertion) (ASI) mechanism, which inserts for you the missing semicolons.

> Helped by ASI, **you can remove** the semicolons from the previous example: [...] The above text is a valid JavaScript code. The missing semicolons are automatically inserted for you.

Also, regarding your suggestion:

> You may want to write another article about semicolons, since this topic, as you rightly mention, is rarely well understood.

Nope, I won't do this. The main reason is that I don't have time. And another reason is that I've written my posts for free, but some people like you start throwing shit with "Tip 7 is plain wrong and misleading." instead of starting with "thanks for the post", and **politely** write a feedback.
